conans/client/graph/graph.py

```diff
@@ -151,6 +151,16 @@ def check_downstream_exists(self, require):
         source_node = dependant.src
         return source_node.check_downstream_exists(down_require)
 
+    def check_loops(self, new_node):
+        if self.ref == new_node.ref:
+            return self
+        if not self.dependants:
+            return
+        assert len(self.dependants) == 1
```

Comment:
Is this because iterating "Depth First"?

Reply:
Yes, this is only called when a new node is added to the graph, that only happens before having diamonds downstream (and then, only 1 dependant always)